cpp/CMakeLists.txt

```diff
@@ -34,7 +34,7 @@ else()
     else()
         set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra")
     endif()
-    set(CMAKE_CXX_FLAGS_DEBUG "-ggdb")
+    set(CMAKE_CXX_FLAGS_DEBUG "-g3 -fsanitize=address -fsanitize=undefined -fsanitize=float-divide-by-zero -fsanitize=float-cast-overflow")
```

Comment:
Does this work with GCC? As in, if you malloc 4 bytes and write 5 bytes to it, is it actually caught?

Reply:
With ASAN (`-fsanitize=address`), yes!

```c
// main.c
#include<stdlib.h>

int main(int argc, char **argv) {
  int * a = malloc(4 * sizeof(int));
  return a[4];
}
```

```bash
gcc -o main main.c -g -fsanitize=address
./main
```
```
=================================================================
==225240==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x602000000020 at pc 0x0000004011ae bp 0x7ffec3e575e0 sp 0x7ffec3e575d8
READ of size 4 at 0x602000000020 thread T0
    #0 0x4011ad in main /tmp/main.c:5
    #1 0x7fdcd6e49b49 in __libc_start_call_main (/lib64/libc.so.6+0x27b49) (BuildId: 245240a31888ad5c11bbc55b18e02d87388f59a9)
    #2 0x7fdcd6e49c0a in __libc_start_main_alias_2 (/lib64/libc.so.6+0x27c0a) (BuildId: 245240a31888ad5c11bbc55b18e02d87388f59a9)
    #3 0x401094 in _start (/tmp/main+0x401094) (BuildId: 57db39438b4b6612eafbc1bf4ef105bb8361ba8c)

0x602000000020 is located 0 bytes after 16-byte region [0x602000000010,0x602000000020)
allocated by thread T0 here:
    #0 0x7fdcd70d92ef in malloc (/lib64/libasan.so.8+0xd92ef) (BuildId: e5f0a0d511a659fbc47bf41072869139cb2db47f)
    #1 0x40116e in main /tmp/main.c:4
    #2 0x7fdcd6e49b49 in __libc_start_call_main (/lib64/libc.so.6+0x27b49) (BuildId: 245240a31888ad5c11bbc55b18e02d87388f59a9)
    #3 0x7fdcd6e49c0a in __libc_start_main_alias_2 (/lib64/libc.so.6+0x27c0a) (BuildId: 245240a31888ad5c11bbc55b18e02d87388f59a9)
    #4 0x401094 in _start (/tmp/main+0x401094) (BuildId: 57db39438b4b6612eafbc1bf4ef105bb8361ba8c)

SUMMARY: AddressSanitizer: heap-buffer-overflow /tmp/main.c:5 in main
Shadow bytes around the buggy address:
  0x601ffffffd80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x601ffffffe00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x601ffffffe80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x601fffffff00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x601fffffff80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x602000000000: fa fa 00 00[fa]fa fa fa fa fa fa fa fa fa fa fa
  0x602000000080: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x602000000100: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x602000000180: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x602000000200: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x602000000280: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==225240==ABORTING
```